Java/parkservices/src/main/java/com/amazon/randomcutforest/extendedrandomcutforest/threshold/BasicThresholder.java

```diff
@@ -0,0 +1,266 @@
+        this.initialThreshold = state.getInitialThreshold();
+        this.lowerThreshold = state.getLowerThreshold();
+        this.zFactor = state.getZFactor();
+        this.upperZfactor = state.getUpperZfactor();
+    }
+
+
+    protected boolean isDeviationReady(){
+        if (count < minimumScores) {
+            return false;
+        }
+
+        if (horizon == 0) {
+            return secondaryDeviation.count >= minimumScores;
+        } else if (horizon == 1) {
+            return primaryDeviation.count >= minimumScores;
+        } else {
+            return secondaryDeviation.count >= minimumScores && primaryDeviation.count >= minimumScores;
+        }
+    }
+
+    protected double intermediateTermFraction(){
+        if (count < minimumScores){
+            return 0;
+        } else if (count> 2*minimumScores){
+            return 1;
+        } else {
+            return (count - minimumScores)*1.0/minimumScores;
+        }
+    }
+
+    protected boolean isLongTermReady(){
+        return (intermediateTermFraction() == 1);
+    }
+
+    protected double basicThreshold(double factor) {
+        if (!isDeviationReady()) { // count < minimumScore is this branch
+            return Math.max(initialThreshold,lowerThreshold);
+        } else if (isLongTermReady()) {
+            return longTermThreshold(factor);
+        } else {
+            return Math.max(lowerThreshold,intermediateTermFraction()*longTermThreshold(factor)
+                    + (1-intermediateTermFraction())*initialThreshold);
+        }
+
+    }
+    protected double longTermThreshold(double factor){
+        return Math.max(lowerThreshold,primaryDeviation.getMean() + factor * longTermDeviation());
+    }
+
+    protected double longTermDeviation(){
+        return ( horizon * primaryDeviation.getDeviation() +
+                (1-horizon) *secondaryDeviation.getDeviation());
+    }
+
+
+    public double getAnomalyGrade(double score,double factor){
+        checkArgument(factor >= zFactor, "incorrect call");
+        // please change here is a first cut
+        if (isLongTermReady()) {
+            if (score < longTermThreshold(factor)) {
+                return 0;
+            }
+            double tFactor = Math.min(upperZfactor, (score - primaryDeviation.getMean()) / longTermDeviation());
+            checkArgument(tFactor >= zFactor, "should not be here");
+            return (tFactor - zFactor) / (upperZfactor - zFactor);
+        } else {
+            if (score <basicThreshold(factor)){
+                return 0;
+            }
+            double upper = Math.max(upperThreshold,2*basicThreshold(factor));
+            double quasiScore = Math.min(score, upper);
+            return (quasiScore - basicThreshold(factor))/(upper - basicThreshold(factor));
+        }
+    }
+
+    public double getAnomalyGrade(double score){
+        return getAnomalyGrade(score,zFactor);
+    }
+
+    public double getConfidenceScore(double score){
```

Comment:
What is this method intended to do?

Reply:
Adding comments. This is the basic deciding logic which provides a "grade" -- current usage is [0:1] where 0 is not an anomaly. It made sense to have this function be the single source of the 0 vs non-zero decision.

Reply:
The Thresholded forest simply decides to invoke/not.

Reply:
Ok - I misread line numbers. Confidence is a score that indicates certainty of the anomaly decision. Typically it is supposed to increase with more observations. I am waiting for these to be filled in -- or I can do it once the definition is finalized.